File: blog/photo.js
const Creator = require("./creator");
const SizedPhoto = require("./sizedPhoto");
const Post = require("./post");
const Moment = require("moment");
const _ = require("lodash");

class Photo extends Post {
	constructor(id, type, source, dateCreated, datePublished, width, height, sizedPhotos, sourceUrl, title, body, creator) {
		super(
			id,
			type || "Photo",
			source,
			dateCreated,
			datePublished,
			title,
			body,
			sourceUrl,
			creator
		);

		let that = this;

		this.width = width;
		this.height = height;
		this.sizedPhotos = [];

		_.each(sizedPhotos, (sizedPhoto) => {
			that.sizedPhoto = sizedPhoto;
		});
	}

	static fromJSON(json) {
		return new Photo(
			json.id,
			json.type,
			json.source,
			json.dateCreated && Moment.utc(json.dateCreated),
			json.datePublished && Moment.utc(json.datePublished),
			json.width,
			json.height,
			json.sizedPhotos.map(SizedPhoto.fromJSON),
			json.sourceUrl,
			json.title,
			json.body,
			json.creator && Creator.fromJSON(json.creator)
		);
	}

	getSizedPhoto(width) {
		const sortedSizedPhotos = _.sortBy(this.sizedPhotos, ["width"]);
		const widthAppropriatePhotos = _.reject(sortedSizedPhotos, (sizedPhoto) => {
			return sizedPhoto.width < width;
		});
		return _.first(widthAppropriatePhotos) || _.last(sortedSizedPhotos);
	}

	set sizedPhoto(sizedPhoto) {
		if (!sizedPhoto.height) {
			sizedPhoto.height = scaleHeightToWidth(sizedPhoto.width, this.width, this.height);
		}
		this.sizedPhotos.push(sizedPhoto);
	}
}

module.exports = Photo;

function scaleHeightToWidth(limitedWidth, originalWidth, originalHeight) {
	return ~~((originalHeight / originalWidth) * limitedWidth);
}

File: blog/photos/photo.py
from blog.post import Post
from blog.creator import Creator


class Photo(Post):
    """

    """

    def __init__(self, type, id, source, title, body, datePublished, dateCreated, sourceUrl, creator, width, height, sizedPhotos):
        Post.__init__(self, type, id, source, title, body, datePublished, dateCreated, sourceUrl, creator)
        self.width = width
        self.height = height
        self.sizedPhotos = sizedPhotos

    @staticmethod
    def fromJSON(json):
        return Photo(
            json.get("type"),
            json.get("id"),
            json.get("source"),
            json.get("title"),
            json.get("body"),
            json.get("datePublished"),
            json.get("dateCreated"),
            json.get("sourceUrl"),
            Creator.fromJSON(json.get("creator")) if json.get("creator") else None,
            json.get("width"),
            json.get("height"),
            json.get("sizedPhotos"),
        )
